src/cargo/util/lints.rs

```diff
@@ -605,6 +608,157 @@ fn output_unknown_lints(
+                        let diag = level.title(&*title).snippet(
+                            Snippet::source(manifest.contents())
+                                .origin(&manifest_path)
+                                .annotation(level.span(span))
+                                .fold(true)
+                        );
+                        gctx.shell().print_message(diag)?;
+                    } else {
+                        let title = format!(
+                            "unexpected `cfg` condition value: {value} for `{cfg}` in `[target.'cfg({cfg_expr})']`"
+                        );
+                        let help_where = format!("occurred in `{path}`", path = path.display());
+                        let diag = level.title(&*title).footer(Level::Help.title(&*help_where));
+                        gctx.shell().print_message(diag)?;
+                    }
+                }
+                None => {
+                    let level = lint_level.to_diagnostic_level();
+                    if lint_level == LintLevel::Forbid || lint_level == LintLevel::Deny {
+                        *error_count += 1;
+                    }
+
+                    let for_cfg = match value {
+                        Some(_value) => Cow::from(format!(" for `{cfg}`")),
+                        None => Cow::from(""),
+                    };
+
+                    // Retrieving the span can fail if our pretty-priting doesn't match what the
+                    // user wrote: `unix="foo"` and `unix = "foo"`, for that reason we don't fail
+                    // and just produce a slightly worth diagnostic.
+                    if let Some(span) = get_span(manifest.document(), &["target", &*format!("cfg({cfg_expr})")], false) {
+                        let manifest_path = rel_cwd_manifest_path(path, gctx);
+                        let title = format!(
+                            "unexpected `cfg` condition name: {name}{for_cfg}"
+                        );
+                        let diag = level.title(&*title).snippet(
+                            Snippet::source(manifest.contents())
+                                .origin(&manifest_path)
+                                .annotation(level.span(span))
+                                .fold(true)
+                        );
+                        gctx.shell().print_message(diag)?;
+                    } else {
+                        let title = format!(
+                            "unexpected `cfg` condition name: {name}{for_cfg} in `[target.'cfg({cfg_expr})']`"
+                        );
+                        let help_where = format!("occurred in `{path}`", path = path.display());
+                        let diag = level.title(&*title).footer(Level::Help.title(&*help_where));
+                        gctx.shell().print_message(diag)?;
+                    }
+                }
+                _ => { /* not unexpected */ }
+            }
+            Ok(())
+        })?;
+    }
```

Comment:
If you use the original manifest's `target` table, any `cfg` will be in its original format, so we can always get a `span`. One thing to note is that it will be the span for the whole `cfg`, not a specific part, which would be ideal. To get the particular part of a `cfg`, we would either need to "find" that part in the `cfg`, or we would need to track spans while parsing `CfgExpr` (I tested a basic implementation of this, and it does work).
```suggestion
    let manifest_path = rel_cwd_manifest_path(path, gctx);

    let Some(target) = manifest.original_toml().target.as_ref() else {
        return Ok(());
    };
    for name in target.keys() {
        if let Some(cfg_expr_orig) = name.strip_prefix("cfg(").and_then(|s| s.strip_suffix(')')) {
            let Ok(cfg_expr) = cfg_expr_orig.parse::<CfgExpr>() else {
                continue;
            };
            cfg_expr.walk(|cfg| -> CargoResult<()> {
                let (name, value) = match cfg {
                    Cfg::Name(name) => (name, None),
                    Cfg::KeyPair(name, value) => (name, Some(value.to_string())),
                };
                let cfg_str = format!("cfg({cfg_expr_orig})");
                if let Some(title) = match check_cfg.expecteds.get(name) {
                    Some(ExpectedValues::Some(values)) if !values.contains(&value) => {
                        let value = match value {
                            Some(value) => Cow::from(format!("`{value}`")),
                            None => Cow::from("(none)"),
                        };
                        Some(format!("unexpected `cfg` condition value: {}", value))
                    }
                    None => Some(format!("unexpected `cfg` condition name: `{}`", name)),
                    _ => None,
                } {
                    if let Some(span) = get_span(manifest.document(), &["target", &cfg_str], false)
                    {
                        let level = lint_level.to_diagnostic_level();
                        if lint_level == LintLevel::Forbid || lint_level == LintLevel::Deny {
                            *error_count += 1;
                        }

                        let diag = level.title(&*title).snippet(
                            Snippet::source(manifest.contents())
                                .origin(&manifest_path)
                                .annotation(level.span(span))
                                .fold(true),
                        );
                        gctx.shell().print_message(diag)?;
                    }
                }
                Ok(())
            })?;
        }
    }
```

Reply:
This looks cleaner than re-using the dependencies, as it avoids duplicate and is more easily forward-compatible. Applied your suggestion. Thanks.

As for getting the span of the particular cfg we want to lint on I think the best way would be to track the spans when parsing the `CfgExpr`/`Cfg` as you already alluded to. I think this would be good follow-up work as this PR is already quite large.